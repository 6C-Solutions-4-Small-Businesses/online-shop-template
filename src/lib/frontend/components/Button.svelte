<script lang="ts">
    import {writable, type Writable} from 'svelte/store'

    export let id: string
    export let classNames: string
    export let background: string = 'bg-white hover:bg-orange-50'
    export let border: string = 'border'
    export let text: string = 'text-black'
    export let disabled: boolean = false
    export let onClick: () => Promise<void>

    const isProcessing: Writable<boolean> = writable(false)

    async function onClickHandler(): Promise<void> {
        isProcessing.set(true)
        await onClick()
        isProcessing.set(false)
    }

</script>

<button
        class={`flex justify-around items-center uppercase rounded-sm font-bold ${classNames} ${disabled ? 'bg-slate-200  border border-black text-black': `${background} ${border} ${text}`}`}
        data-testid="{id}"
        disabled={$isProcessing || disabled}
        id="{id}"
        on:click={onClickHandler}
>
    {#if ($isProcessing)}
        <svg
                class="animate-spin"
                width="24"
                height="24"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
                stroke-width="2"
                stroke-linecap="round"
        >
            <circle
                    cx="12"
                    cy="12"
                    r="10"
                    fill="none"
                    stroke="#FFA500"
                    stroke-dasharray="30"
                    stroke-dashoffset="0"
            >
            </circle>
        </svg>
    {:else }
        <slot/>
    {/if}
</button>
