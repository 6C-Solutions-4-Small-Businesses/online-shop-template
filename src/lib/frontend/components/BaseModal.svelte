<script lang="ts">
    import Button from '$lib/frontend/components/Button.svelte'
    import type {SvelteComponent} from 'svelte'
    import CloseCircleIcon from '~icons/mdi/close-circle'

    export let title: string | undefined
    export let body: string | undefined
    export let isHtml: boolean = false
    export let isFull = false
    export let background = 'bg-white'
    export let parent: SvelteComponent
    export let buttonTextCancel: string | undefined = undefined
    export let buttonTextSubmit: string | undefined = undefined
    export let submitButtonDisabled = false
    export let onSubmitClickHandler: () => Promise<void>
    export let onCancelClickHandler: (() => void) | undefined = undefined
</script>

<div class={`${background} ${isFull ? 'w-screen h-screen overflow-scroll pb-5' : 'w-full'} flex flex-col gap-7 items-center rounded-md`}>
    <header
            class=" w-full flex  justify-between items-center text-md font-bold uppercase bg-green-50 h-14 text-center p-3 text-black rounded-t-md">
        <p>{title }</p>
        <button class="w-8 h-8 flex items-center justify-center rounded-full" data-testid="close-modal-button"
                on:click={onCancelClickHandler ?? parent.onClose}>
            <CloseCircleIcon/>
        </button>
    </header>
    {#if isHtml}
        <div class="px-3 text-slate-600">{@html body}</div>
    {:else if body}
        <p class="w-11/12 text-start text-slate-600">{body}</p>
    {/if}
    <slot/>
    {#if buttonTextCancel || buttonTextSubmit}
        <footer class="w-11/12 flex flex-row md:justify-between gap-2 mb-5">
            {#if buttonTextCancel}
                <Button
                        background="bg-white hover:bg-orange-50"
                        classNames="w-full md:w-2/3 btn {parent.buttonNeutral}  h-12"
                        id="cancel-button"
                        onClick={onCancelClickHandler ?? parent.onClose}>
                    {buttonTextCancel}
                </Button>
            {/if}
            {#if buttonTextSubmit}
                <Button
                        background="bg-orange-50 hover:bg-white"
                        border="border border-orange-500"
                        classNames="w-full md:w-2/3 shadow-sm h-12 btn {parent.buttonPositive}"
                        disabled={submitButtonDisabled}
                        id="continue-button"
                        onClick={onSubmitClickHandler}
                        text="text-orange-500"
                >
                    {buttonTextSubmit}
                </Button>
            {/if}
        </footer>
    {/if}
</div>
