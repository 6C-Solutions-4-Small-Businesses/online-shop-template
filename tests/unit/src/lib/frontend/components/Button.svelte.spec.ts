import {afterEach, beforeEach, describe, expect} from 'vitest'
import '@testing-library/jest-dom'
import {render, type RenderResult, screen as vScreen, waitFor} from '@testing-library/svelte'
import BaseModal from '$lib/frontend/components/BaseModal.svelte'
import type {SvelteComponent} from 'svelte'
import {mock} from 'vitest-mock-extended'
import Button from '$lib/frontend/components/Button.svelte'

const buttonId = 'button-id'
describe('Button component', () => {
    const title = 'Title'
    const body = 'Body'
    const buttonTextCancel = 'annuler'
    const buttonTextSubmit = 'soumettre'
    const parentOnModalClosed = vi.fn()
    const onClickHandler = vi.fn()

    let view: RenderResult<Button, typeof import('@testing-library/dom/types/queries')>

    beforeEach(() => {
        view = render(Button, {
            props: {
                id: buttonId,
                text: 'Button text',
                onClick: onClickHandler,
                border: 'border-2',
                background: 'bg-blue-500',
                block: true,
                disabled: false,
                classNames: 'text-white',
            }
        })
    })

    describe('Structure', () => {
        it('should have match snapshot', () => {
            expect(view.container).toMatchSnapshot()
        })
    })

    describe('Behavior', () => {
        it('should call "onClickHandler" when button is clicked', async () => {
            const button = vScreen.getByTestId(buttonId)
            button.click()

            await waitFor(() => expect(onClickHandler).toHaveBeenCalled())
        })
    })

    afterEach(() => {
        view.unmount()
        vi.clearAllMocks()
    })
})
