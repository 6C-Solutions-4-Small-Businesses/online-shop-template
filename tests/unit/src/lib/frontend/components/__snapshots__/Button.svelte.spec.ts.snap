// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > Structure > should have match snapshot 1`] = `
<body>
  <div>
    <button
      class=" justify-around items-center uppercase rounded-sm font-bold text-white bg-blue-500 border-2 Button text"
      data-testid="button-id"
      id="button-id"
    />
    <!--&lt;Button&gt;-->
  </div>
</body>
`;
